packages/x-data-grid/src/components/GridSkeletonLoadingOverlay.tsx

```diff
@@ -0,0 +1,135 @@
+import * as React from 'react';
+import Skeleton from '@mui/material/Skeleton';
+
+import { styled } from '@mui/system';
+import { unstable_useForkRef as useForkRef } from '@mui/utils';
+import { useGridApiContext } from '../hooks/utils/useGridApiContext';
+import { gridColumnPositionsSelector, gridColumnsTotalWidthSelector } from '../hooks';
+import { GridColType, GridEventListener } from '../models';
+import { seededRandomNumberGenerator } from '../utils/utils';
+
+const DEFAULT_COLUMN_WIDTH_RANGE = [40, 80] as const;
+
+const COLUMN_WIDTH_RANGE_BY_TYPE: Partial<Record<GridColType, [number, number]>> = {
+  number: [40, 60],
+  string: [40, 80],
+  date: [40, 60],
+  dateTime: [60, 80],
+  singleSelect: [40, 80],
+} as const;
+
+const colWidthVar = (index: number) => `--colWidth-${index}`;
+
+const SkeletonOverlay = styled('div')({
+  display: 'grid',
+  overflow: 'hidden',
+});
+
+const SkeletonCell = styled('div')({
+  display: 'flex',
+  flexDirection: 'row',
+  alignItems: 'center',
+  borderBottom: '1px solid var(--DataGrid-rowBorderColor)',
+});
+
+const GridSkeletonLoadingOverlay = React.forwardRef<
+  HTMLDivElement,
+  React.HTMLAttributes<HTMLDivElement>
+>(function GridSkeletonLoadingOverlay(props, forwardedRef) {
+  const ref = React.useRef<HTMLDivElement>(null);
+  const handleRef = useForkRef(ref, forwardedRef);
+
+  const apiRef = useGridApiContext();
+
+  const dimensions = apiRef.current?.getRootDimensions();
+  const viewportHeight = dimensions?.viewportInnerSize.height ?? 0;
+
+  // @ts-expect-error Function signature expects to be called with parameters, but the implementation suggests otherwise
+  const rowHeight = apiRef.current.unstable_getRowHeight();
+  const skeletonRowsCount = Math.ceil(viewportHeight / rowHeight);
+
+  const totalWidth = gridColumnsTotalWidthSelector(apiRef);
+  const positions = gridColumnPositionsSelector(apiRef);
```

Comment:
In render code you'll want to use `useGridSelector(apiRef, selector)`, never the raw selector. The hook will re-render when the selector changes. You'll also want to avoid `apiRef.current.getSomething()`, you should use the equivalent selector with the hook.

Reply:
Thanks, fixed here https://github.com/mui/mui-x/pull/13293/commits/d32fa8b1c45257fd054d67b96745f962686502b5